docs/man/man1/hbmgui.1
.\" Copyright (c) 2020 andypugh
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
.\" USA.
.\"
.\"
.\"
.TH hbmgui "1"  "2020-08-26" "LinuxCNC Documentation" "The Enhanced Machine Controller"
.SH NAME
hbmgui \- Vismach Virtual Machine GUI

.SH DESCRIPTION
\fBhbmgui\fR is one of the sample \fBVismach\fR GUIs for LinuxCNC,
simulating a Horizontal Boring Machine.

See the main LinuxCNC documentation for more details. 

http://linuxcnc.org/docs/html/gui/vismach.html
.
.SH "SEE ALSO"
\fBLinuxCNC(1)\fR

Much more information about LinuxCNC and HAL is available in the LinuxCNC
and HAL User Manuals, found at /usr/share/doc/LinuxCNC/.

.SH BUGS
None known at this time. 
.PP
.SH AUTHOR
This man page written by andypugh, as part of the LinuxCNC project.
.SH REPORTING BUGS
Report bugs at https://github.com/LinuxCNC/linuxcnc/issues
.SH COPYRIGHT
Copyright \(co 2020 andypugh.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


docs/NEWS
check debian/changelog for recent news

# some of EMC2 History & News
#############################

2009.06.13 - mshaver added emc2/configs/smithy directory and related files
             for Smithy Company's line of cnc machines.

2006.11.17 - removed HAL types s8, u8, s16, and u16.  HAL now uses float
	     for analog values, bit for booleans, and s32 or u32 for
	     integer values.  The only users who will be affected are
	     those with Universal Stepper Controller cards.  See item 9 at
	     http://wiki.linuxcnc.org/cgi-bin/emcinfo.pl?UpdatingConfigurationsForDevelopmentVersions

2006.11.15 - merged classicladder 0.7.100

2006.11.04 - added a feed-hold HAL pin.  Added M codes to enable/disable
	     feed override, spindle override, feed hold, and adaptive feed
	     individually.  Feed override is M50, spindle override is M51,
	     adaptive feed is M52, and feed hold is M53.  G50/51 are no
	     longer used to enable/disable adaptive feed, use M52 instead.

2006.10.24 - added support for block delete as defined by Section 2.2.2
             in the RS274NGC_3 specs.

2006.10.24 - added support for optional program stop as defined by Section 
             2.2.3 in the RS274NGC_3 specs.

2006.10.14 - manpages have been created for most rtapi and hal calls

2006.10.01 - added a userspace simulator which includes rtapi threads.
	     Use "--enable-simulator" to enable this mode, which cannot
	     be used to control real hardware

2006.09.27 - added support to optionally save joint positions between
             runs ([TRAJ] POSITION_FILE = position.txt)

2006.09.26 - added support to specify UNITS in literal in the ini. 
             (e.g. [TRAJ] LINEAR_UNITS = mm )

2006.09.09 - fixed bug #1384883, optional stop was not optional.

2006.08.09 - AXIS is now a part of EMC2, not a separate source download.

2006.08.07 - Added a spindle-speed override. This allows to override the spindle
	     speed during a program run. The MIN_SPINDLE_OVERRIDE and 
	     MAX_SPINDLE_OVERRIDE in the ini assure that speed doesn't get 
	     reduced or increased too much. Both values are in %.

2006.08.06 - Documentation can be built from source with the 'make docs' target

2006.06.11 - added realtime support for jogwheels.  Three HAL pins per axis:
	     one accepts raw counts from the wheel, one sets the distance moved
	     per count, and one enables that axis to move.

2006.05.17 - added 'adaptive feed', a HAL pin that can be used to scale the
	     feedrate in real time.  Simular to feedrate override, but can
	     only vary from 0.0 to 1.0, and reacts much faster.  Intended 
	     for use with EDM and other applications that require closed 
	     loop control of feedrate.

2006.03.18 - A new pin "count" exists for freqgen.  It is the same as the
             internal symbol "count" shown in the hal introduction.pdf,
             or the result of up OR down if there was an "OR" component.

2006.03.16 - Modified the motion control module to behave more like other
             HAL modules - it no longer adds its functions to threads
             automatically.  All the sample configs have been changed to
             explicitly add the functions, users who have custom configs
             will need to do the same.

2006.03.12 - If a copy of axis 1.3a0 or newer is detected inside src/
             it will be automatically built and installed for
             --enable-run-in-place use

2006.03.10 - minor bugfix in tkemc for jogging

2006.03.10 - added genhexkins, with a minitetra sample config.
             This proves EMC2 works just as good with hexapods as EMC1 did.

2006.03.10 - added VTI support & sample configs

2006.03.07 - lots of fixes to the TP (most were wrong in the old one as well)

2006.03.05 - new Trajectory Planner

2006.02.20 - added a manpage for the runscript.

2006.02.12 - Fixed a bug that was causing the RT code to run very slowly,
             forcing users to set BASE_PERIOD very long.

2006.02.10 - The build process has been replaced, no more recursive $(MAKE)

2006.01.28 - fixed bug #1205237, mode switching sometimes hung

2006.01.08 - M62/M63 digital outputs synchronous to motion
             M64/M65 digital outputs that get set right away

2006.01.06 - work on halconfig.tcl, it is used to display/configure HAL data
             requires the BWidget package, realtime, and HAL to run.

2006.01.04 - fix bug #1387740, a motion that goes nowhere clears feedrate

2006.01.01 - the config picker has i18n support, a German translation exists

2005.12.20 - a new config picker has been written, if you start just the
             runscript this one helps you choose a config to run

2005.12.13 - new runscript, and config organization
             work towards the first release of EMC2

2005.xx.xx - EMC2 now supports tool-changing in complex scenarios

2005.xx.xx - classicladder now a HAL component, a software PLC

2004.xx.xx - HAL (the Hardware Abstraction Layer) is used for
	     addressing hardware drivers, and allows for simple configuration


#developers: this file should hold a history of improvements, bugfixes, etc
             it is useful for keeping track of development


src/hal/user_comps/vismach/hbmgui.py
#!/usr/bin/env python3
#
#    Visulization model of a Horizontal Boring Mill with quill
#
#    Copyright 2007 John Kasunich
#    Derived from a work by John Kasunich, Jeff Epler, and Chris Radek
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


from vismach import *
import hal
import math
import sys

# distance from spindle nose to tool-length = 0 point
tool_length_offset = 4.0

# give endpoint Z values and radii
# resulting cylinder is on the Z axis
class HalToolCylinder(CylinderZ):
    def __init__(self, comp, *args):
        CylinderZ.__init__(self, *args)
        self.comp = comp

    def coords(self):
        return (0, self.comp["tool-radius"],
        -tool_length_offset-self.comp["tool-length"], self.comp["tool-radius"])


c = hal.component("hbmgui")
# HAL pins
c.newpin("table", hal.HAL_FLOAT, hal.HAL_IN)
c.newpin("saddle", hal.HAL_FLOAT, hal.HAL_IN)
c.newpin("head", hal.HAL_FLOAT, hal.HAL_IN)
c.newpin("quill", hal.HAL_FLOAT, hal.HAL_IN)

c.newpin("tool-length", hal.HAL_FLOAT, hal.HAL_IN)
c.newpin("tool-radius", hal.HAL_FLOAT, hal.HAL_IN)
c.ready()

tool_radius=0.5

for setting in sys.argv[1:]: exec(setting)

# the approach here is a little different than in previous vismach models
# we define the geometry of all parts first
# only at the end to we build the kinematic chain(s)

# floor
floor = Collection([
        # a big slab
        Color([0.2,0.2,0.2,0], [Box(-150, -4, -150, 150, 0, 150)])
        ])

# saddle ways
saddle_way_width = 5.0
saddle_way_spacing = 72.0
saddle_way_start = -50
saddle_way_end = 50
saddle_slide_length = 48
column_center_x = 24
column_center_z = 75
column_width = 24
column_depth = 30
column_height = 144
table_thick = 8
table_length = 120
table_width = 60
table_way_width = 5.0
column_way_top_x = column_center_x-0.5*column_width-1.0
head_width = 2*(column_way_top_x-1.0)
head_top = 15
head_bottom = -9
head_depth = 36
quill_length = 48
quill_dia = 6.0
column_head_offset = 4.0

saddleways = Collection([
        Box(-1.5*saddle_way_spacing-0.5*saddle_way_width, -1.0, saddle_way_start,
            -1.5*saddle_way_spacing+0.5*saddle_way_width, 1.0, saddle_way_end),
        Box( 1.5*saddle_way_spacing-0.5*saddle_way_width, -1.0, saddle_way_start,
             1.5*saddle_way_spacing+0.5*saddle_way_width, 1.0, saddle_way_end),
        Box(-0.5*saddle_way_spacing-0.5*saddle_way_width,  8.0, saddle_way_start,
            -0.5*saddle_way_spacing+0.5*saddle_way_width, 10.0, saddle_way_end),
        Box( 0.5*saddle_way_spacing-0.5*saddle_way_width,  8.0, saddle_way_start,
             0.5*saddle_way_spacing+0.5*saddle_way_width, 10.0, saddle_way_end),
        # and screw
        Translate([CylinderZ(saddle_way_start+8, 1.5, saddle_way_end-7, 1.5)], 0, 8, 0)
        ])
saddleways = Color([0.5,0.5,0.8,0],[saddleways])

base = Collection([
        # section under each saddle way
        Box(-0.5*saddle_way_spacing-0.4*saddle_way_width, 5.0, saddle_way_start+1,
            -0.5*saddle_way_spacing+0.8*saddle_way_width, 9.0, saddle_way_end-1),
        Box( 0.5*saddle_way_spacing-0.8*saddle_way_width, 5.0, saddle_way_start+1,
             0.5*saddle_way_spacing+0.4*saddle_way_width, 9.0, saddle_way_end-1),
        # main portion connecting saddle ways
        Box(-0.5*saddle_way_spacing-0.5*saddle_way_width, 0.0, saddle_way_start+1,
             0.5*saddle_way_spacing+0.5*saddle_way_width, 5.0, saddle_way_end-1),
        # screw mount blocks
        Box( -3, 5, saddle_way_end-7, 3, 10.5, saddle_way_end-1),
        Box( -3, 5, saddle_way_start+8, 3, 11.0, saddle_way_start),
        # motor
        Translate([CylinderZ(saddle_way_start, 3.5, saddle_way_start-18, 3.5)], 0, 8, 0),
        # portion under column
        Box((column_center_x-column_width*1.25), 0.0, saddle_way_end-1,
             (column_center_x+column_width*0.75), 5.0, column_center_z+column_depth*0.75)
        ])
base = Color([0.5,0.9,0.9,0],[base])


saddle_slides = Collection([
        Box(-0.5*saddle_way_spacing-0.8*saddle_way_width,  8.5, -0.5*saddle_slide_length,
            -0.5*saddle_way_spacing+0.4*saddle_way_width, 11.5,  0.5*saddle_slide_length),
        Box( 0.5*saddle_way_spacing-0.4*saddle_way_width,  8.5, -0.5*saddle_slide_length,
             0.5*saddle_way_spacing+0.8*saddle_way_width, 11.5,  0.5*saddle_slide_length),
        Box(-1.5*saddle_way_spacing-0.4*saddle_way_width, 1.0, -0.5*saddle_slide_length,
            -1.5*saddle_way_spacing+0.4*saddle_way_width, 2.5,  0.5*saddle_slide_length),
        Box( 1.5*saddle_way_spacing-0.4*saddle_way_width, 1.0, -0.5*saddle_slide_length,
             1.5*saddle_way_spacing+0.4*saddle_way_width, 2.5,  0.5*saddle_slide_length),
        # and nut
        Translate([CylinderZ(-0.5*saddle_slide_length+10, 2.2, 0.5*saddle_slide_length-10, 2.2)], 0, 8, 0)
        ])
saddle_slides = Color([0.7,0.9,0.9,0],[saddle_slides])

saddle = Collection([
        # section riding on end slides
        Box(-1.5*saddle_way_spacing-0.6*saddle_way_width,  2.5, -0.5*saddle_slide_length+1,
            -1.5*saddle_way_spacing+0.6*saddle_way_width, 11.5,  0.5*saddle_slide_length-1),
        Box( 1.5*saddle_way_spacing-0.6*saddle_way_width,  2.5, -0.5*saddle_slide_length+1,
             1.5*saddle_way_spacing+0.6*saddle_way_width, 11.5,  0.5*saddle_slide_length-1),
        # main portion
        Box(-1.5*saddle_way_spacing-0.6*saddle_way_width, 11.5, -0.5*saddle_slide_length+0.5,
             1.5*saddle_way_spacing+0.6*saddle_way_width, 15.0,  0.5*saddle_slide_length-0.5),
        Box(-1.5*saddle_way_spacing-0.6*saddle_way_width, 15.0, -0.5*saddle_slide_length+0.5,
             1.5*saddle_way_spacing+0.6*saddle_way_width, 19.0, -0.5*saddle_slide_length+5.5),
        Box(-1.5*saddle_way_spacing-0.6*saddle_way_width, 15.0,  0.5*saddle_slide_length-0.5,
             1.5*saddle_way_spacing+0.6*saddle_way_width, 19.0,  0.5*saddle_slide_length-5.5),
        # nut block
        Box(-5, 5.5, -0.5*saddle_slide_length+11, 5, 11.5,  0.5*saddle_slide_length-11),
        # screw mount blocks
        Box(-1.5*saddle_way_spacing+20, 15.0, -3,
            -1.5*saddle_way_spacing+16, 20.5,  3),
        Box( 1.5*saddle_way_spacing-5, 15.0, -3,
             1.5*saddle_way_spacing+5, 20.5,  3),
        # motor
        Translate([CylinderX(1.5*saddle_way_spacing+5, 3.5, 1.5*saddle_way_spacing+25, 3.5)], 0, 18, 0),
        ])
saddle = Color([0.5,0.9,0.9,0],[saddle])


table_ways = Collection([
        Box(-1.5*saddle_way_spacing-0.7*saddle_way_width, 18.0, -0.5*saddle_slide_length,
             1.5*saddle_way_spacing+0.7*saddle_way_width, 20.0, -0.5*saddle_slide_length+table_way_width),
        Box(-1.5*saddle_way_spacing-0.7*saddle_way_width, 18.0,  0.5*saddle_slide_length,
             1.5*saddle_way_spacing+0.7*saddle_way_width, 20.0,  0.5*saddle_slide_length-table_way_width),
        # and screw
        Translate([CylinderX(-1.5*saddle_way_spacing+20, 1.5, 1.5*saddle_way_spacing+10, 1.5)], 0, 18, 0)
        ])
table_ways = Color([0.5,0.5,0.8,0],[table_ways])


table_slides = Collection([
        Box(-0.5*table_length, -table_thick+1.0, -0.5*saddle_slide_length-0.3*table_way_width,
             0.5*table_length, -table_thick-1.5, -0.5*saddle_slide_length+0.9*table_way_width),
        Box(-0.5*table_length, -table_thick+1.0,  0.5*saddle_slide_length+0.3*table_way_width,
             0.5*table_length, -table_thick-1.5,  0.5*saddle_slide_length-0.9*table_way_width),
        # and nut
        Translate([CylinderX(-0.5*table_length+20, 2.2, 0.5*table_length-15, 2.2)], 0, -table_thick-2, 0)
        ])
table_slides = Color([0.7,0.9,0.9,0],[table_slides])

# generate the t-slots with a loop
w12 = table_width/12
table_top = []
for s in range(-4, 5, 2):
        table_top = table_top + [
                Box(-0.5*table_length, -2, (s-1)*w12+1.0,
                     0.5*table_length, -1, (s+1)*w12-1.0),
                Box(-0.5*table_length, -1, (s-1)*w12+0.5,
                     0.5*table_length,  0, (s+1)*w12-0.5)]
# end sections are different
s = -6
table_top = table_top + [
        Box(-0.5*table_length, -2,  s*w12,
             0.5*table_length, -1, (s+1)*w12-1.0),
        Box(-0.5*table_length, -1,  s*w12,
             0.5*table_length,  0, (s+1)*w12-0.5)]
s = 6
table_top = table_top + [
        Box(-0.5*table_length, -2, (s-1)*w12+1.0,
             0.5*table_length, -1,  s*w12),
        Box(-0.5*table_length, -1, (s-1)*w12+0.5,
             0.5*table_length,  0,  s*w12)]


table = Collection(table_top +[
        # solid part
        Box(-0.5*table_length, -2, -0.5*table_width,
             0.5*table_length, -table_thick+1.0,  0.5*table_width),
        # nut block
        Box(-0.5*table_length+21, -table_thick-4.7, -5, 0.5*table_length-16, -table_thick+1.0,  5),
        ])
table = Color([0.5,0.9,0.9,0],[table])

column = Collection([
        # main body
        Box(column_way_top_x+4.5, 5.0, column_center_z-0.5*column_depth,
            column_center_x+0.5*column_width, column_height, column_center_z+0.5*column_depth),
        # way supports
        Box(column_way_top_x+4.5, 5.0, column_center_z-0.5*column_depth,
            column_way_top_x+1.0, column_height, column_center_z-0.5*column_depth+5.0),
        Box(column_way_top_x+4.5, 5.0, column_center_z+0.5*column_depth,
            column_way_top_x+1.0, column_height, column_center_z+0.5*column_depth-5.0),
        # screw/motor mount
        Box(column_way_top_x+1.5-2.5, column_height-5.0, column_center_z-3.0,
            column_way_top_x+1.5+3.0, column_height+1.0, column_center_z+3.0),
        # motor
        Translate([CylinderY(column_height+1.0, 3.5, column_height+19.0, 3.5)], column_way_top_x+1.5, 0, column_center_z),
        ])
column = Color([0.5,0.9,0.9,0],[column])

column_ways = Collection([
        Box(column_way_top_x+2.0, 6.0, column_center_z-0.5*column_depth-1.0,
            column_way_top_x+0.0, column_height+0.5, column_center_z-0.5*column_depth+4.0),
        Box(column_way_top_x+2.0, 6.0, column_center_z+0.5*column_depth+1.0,
            column_way_top_x+0.0, column_height+0.5, column_center_z+0.5*column_depth-4.0),
        # and screw
        Translate([CylinderY(5.0, 1.5, column_height-4.0, 1.5)], column_way_top_x+1.5, 0, column_center_z)
        ])
column_ways = Color([0.5,0.5,0.8,0],[column_ways])


work = Capture()

tooltip = Capture()

tool = Collection([
        Translate([HalTranslate([tooltip], c, "tool-length", 0, 0, -1)], 0, 0, -tool_length_offset),
        HalToolCylinder(c)
        ])

tool = Color([0.8,0.6,0.9,0], [tool] )



# quill - nose at Z = 0, tool points along -Z
quill = Collection([
        tool,
        # main quill body
        CylinderZ(0.0, 0.5*quill_dia, quill_length, 0.5*quill_dia)
        ])
quill = Color([0.8,0.1,0.1,0],[quill])

# head assy - same coords as fully retracted quill
head = Collection([
        # main body
        Box(-0.5*head_width, head_top, 2,
             0.5*head_width, head_bottom, head_depth+2),
        # nut block
        Box(0.5*head_width, head_top-6.0, column_head_offset+0.5*column_depth-3.0,
            column_way_top_x+1.5+2.5, head_bottom+6.0, column_head_offset+0.5*column_depth+3.0),
        # quill housing
        CylinderZ(1, 6, quill_length*1.5, 6)
        ])
head = Color([0.5,0.9,0.9,0],[head])

head_slides = Collection([
        Box(column_way_top_x+1.5, head_top+1.0, -2.5+column_head_offset,
            0.5*head_width, head_bottom-1.0, 3.5+column_head_offset),
        Box(column_way_top_x+1.5, head_top+1.0, column_depth+2.5+column_head_offset,
            0.5*head_width, head_bottom-1.0, column_depth-3.5+column_head_offset),
        # and nut
        Translate([CylinderY(head_top-5.5, 2.2, head_bottom+5.5, 2.2)], column_way_top_x+1.5, 0, column_head_offset+0.5*column_depth)
        ])
head_slides = Color([0.7,0.9,0.9,0],[head_slides])


# kinematic chain
# kins_base -> kins_saddle -> kins_table -> kins_work
# kins_base -> kins_head -> kins_quill

# XY origin at quill centerline, Z = 0 at spindle nose
kins_quill = Collection([
        quill
        ])

# XY origin on quill centerline, Z = 0 at nose of fully retracted quill
kins_head = Collection([
        head,
        head_slides,
        HalTranslate([kins_quill],c,"quill",0,0,1)
        ])

# origin at center of top surface of rotary table (when we eventually add it)
kins_work = Collection([
        #rotary_top,
        work
        ])

# XZ origin at center of top of table
kins_table = Collection([
        table_slides,
        table,
        kins_work,
        ])

# origin at center of top of table ways
kins_saddle = Collection([
        saddle,
        saddle_slides,
        table_ways,
        Translate([HalTranslate([kins_table],c,"table",-1,0,0)], 0, table_thick+20, 0)
        ])

# origin at center of saddle ways, at floor level
kins_base = Collection([
        floor,
        base,
        column,
        column_ways,
        saddleways,
        Translate([HalTranslate([kins_head],c,"head",0,1,0)],0,20+table_thick,column_center_z-0.5*column_depth-column_head_offset),
        Translate([HalTranslate([kins_saddle],c,"saddle",0,0,-1)],0,0,0)
        ])

model = Collection([kins_base])

x_axis = Collection([
        CylinderX(0, 3, 30, 3),
        CylinderX(30,6, 40, 0),
        Box(27, -6, -6, 30, 6, 6)
        ])
x_axis = Color([1,0,0,0],[x_axis])

y_axis = Collection([
        CylinderY(0, 3, 30, 3),
        CylinderY(30,6, 40, 0),
        Box(-6, 27, -6, 6, 30, 6)
        ])
y_axis = Color([0,1,0,0],[y_axis])

z_axis = Collection([
        CylinderZ(0, 3, 30, 3),
        CylinderZ(30,6, 40, 0),
        Box(-6, -6, 27, 6, 6, 30)
        ])
z_axis = Color([0,0,1,0],[z_axis])

axes = Collection([
        Box(-5, -5, -5, 5, 5, 5),
        x_axis,
        y_axis,
        z_axis,
        tooltip,work
        ])

model = Rotate([model],90,1,0,0)


main(model, tooltip, work, 200, lat=-60, lon=60)
